logicrunner/handle_executorresults.go

```diff
@@ -0,0 +1,201 @@
+				queueFromMessage,
+				ExecutionQueueElement{
+					ctx:     qe.Parcel.Context(context.Background()),
+					parcel:  qe.Parcel,
+					request: qe.Request,
+				})
+		}
+		es.Queue = append(queueFromMessage, es.Queue...)
+	}
+
+	es.Unlock()
+
+	return nil
+}
+
+// ---------------
+
+type HandleExecutorResults struct {
+	dep *Dependencies
+
+	Message bus.Message
+}
+
+func (h *HandleExecutorResults) realPresent(ctx context.Context, f flow.Flow) error {
+	parcel := h.Message.Parcel
+	logger := inslogger.FromContext(ctx)
+	msg := parcel.Message().(*message.ExecutorResults)
+
+	// now we have 2 different types of data in message.HandleExecutorResultsMessage
+	// one part of it is about consensus
+	// another one is about prepare state on new executor after pulse
+	// TODO make it in different goroutines
+
+	// prepare state after previous executor
+	procInitializeExecutionState := initializeExecutionState{
+		LR:  h.dep.lr,
+		msg: msg,
+	}
+	if err := f.Procedure(ctx, &procInitializeExecutionState, true); err != nil {
+		if err == flow.ErrCancelled {
+			return nil
+		}
+		err := errors.Wrap(err, "[ HandleExecutorResults ] Failed to initialize execution state")
+		return err
+	}
+
+	if !procInitializeExecutionState.Result.clarifyPending {
+		procClarifyPending := ClarifyPendingState{
+			es:     procInitializeExecutionState.Result.es,
+			parcel: parcel,
+		}
+		procClarifyPending.Dep.ArtifactManager = h.dep.lr.ArtifactManager
+
+		if err := f.Procedure(ctx, &procClarifyPending, true); err != nil {
+			if err == flow.ErrCancelled {
+				return nil
+			}
+
+			err := errors.Wrap(err, "[ HandleExecutorResults ] Failed to clarify pending")
+			return err
+		}
+	}
+
+	ref := msg.GetReference()
+	procStartQueueProcessorIfNeeded := StartQueueProcessorIfNeeded{
+		es:  procInitializeExecutionState.Result.es,
+		dep: h.dep,
+		ref: &ref,
+	}
+	if err := f.Handle(ctx, procStartQueueProcessorIfNeeded.Present); err != nil {
+		if err == flow.ErrCancelled {
+			return nil
+		}
+
+		logger.Warn("[ HandleExecutorResults ] StartQueueProcessorIfNeeded returns error: ", err)
+		return errors.Wrap(err, "[ HandleExecutorResults ] Failed to process queue")
+	}
+
+	return nil
+}
+
+func (h *HandleExecutorResults) Present(ctx context.Context, f flow.Flow) error {
+	parcel := h.Message.Parcel
+	ctx = loggerWithTargetID(ctx, parcel)
+	logger := inslogger.FromContext(ctx)
+
+	logger.Debug("HandleExecutorResults.Present starts ...")
+
+	msg, ok := parcel.Message().(*message.ExecutorResults)
+	if !ok {
+		return errors.New("HandleExecutorResults( ! message.ExecutorResults )")
+	}
+
+	ctx, span := instracer.StartSpan(ctx, "HandleExecutorResults.Present")
+	span.AddAttributes(trace.StringAttribute("msg.Type", msg.Type().String()))
+	defer span.End()
+
+	err := h.realPresent(ctx, f)
+
+	if err != nil {
+		h.Message.ReplyTo <- bus.Reply{Reply: &reply.Error{}, Err: err}
+	} else {
+		h.Message.ReplyTo <- bus.Reply{Reply: &reply.OK{}, Err: nil}
+	}
```

Comment:
You can simplify this - err changes only Reply, Err is always err.

Reply:
rewritten